### PR TITLE
chore(SRVKP-4528): alert/panels pac deadlock/performance

### PR DESCRIPTION
**Pull request by divyansh42**
These are the alerts, panels, and alert unit tests for determining if tekton pac controller is either deadlocked or experiencing abnormally poor performance.

### CODE REVIEW

**Comment by divyansh42** [comment]
@gabemontero please take a look at the PR mostly for the dashboard.
For alerts and tests few changes might be required, that we can discuss async

**Comment by gabemontero** [comment]
@divyansh42 this repo has some rules around git commits 

1) you'll need the change the commit and PR title to have `chore` in it, like

`chore(SRVKP-4528): alert/panels pac deadlock/performance`

2) you did sign the commit - good job there

Also, for just the PR title, add the prefix `WIP:` so that it is clear you are not yet ready for review until we sort out the alerts and their unit tests.

I still need to do my review of the specific changes

File: rhobs/alerting/data_plane/prometheus.pipeline_alerts.yaml
Comment (gabemontero) [inline rhobs/alerting/data_plane/prometheus.pipeline_alerts.yaml:184]: fyi there is probably a lot of overlap with chains wrt this alert ... I'll keep this mind mind as I start working on `chains-performance.md`

that said, as I'm working on these SOPs, the notion of high latency here may not be "actionable" ...related to the Aug 20 konflux arch call and my comment in https://gitlab.cee.redhat.com/konflux/docs/sop/-/merge_requests/163#note_12355622

A possible outcome is we remove this alert and the analogous chains one, at least for the time being

File: test/promql/tests/data_plane/pipeline_deadlock_crashloop_test.yaml
Comment (gabemontero) [inline test/promql/tests/data_plane/pipeline_deadlock_crashloop_test.yaml:304]: before we talk about this stuff voice to voice give your prior comment about being uncertain about this stuff, I'll check out your branch and try to run this locally, see how close you got .... if only minor changes are needed, I'll make suggestions in a separate review.

**Review by gabemontero — CHANGES_REQUESTED** [review]
I am pretty sure we talked about this at one point, but I suspect I only mentioned it briefly, and it got lost in the shuffle.  And as there isn't a specific version for chains, those directions from me would have skipped this.

But if you look at the query for repeated restarts at https://github.com/redhat-appstudio/o11y/blob/beee5d1ad74a1db6de2250a5e32096f38ee7e052/rhobs/alerting/data_plane/prometheus.pipeline_alerts.yaml#L52 

you'll see that `pod=~"tekton-.*"` will not cover any of the pipelines-as-code pods, as they start with `pipelines-as-code..`    Don't worry about the `pac_..` ones or any other of the pods that do not start with `tekton-...`

Either see if you can modify the regular expression `pod=~tekton-.*` to also capture pipelines-as-code, or create a separate alert for PAC that does `pod=~pipelines-as-code-.*`

File: test/promql/tests/data_plane/pipeline_deadlock_crashloop_test.yaml
Comment (gabemontero) [inline test/promql/tests/data_plane/pipeline_deadlock_crashloop_test.yaml:311]:
```suggestion
      - series: 'pac_watcher_work_queue_depth{source_cluster="cluster02"}'
```

You missed a nuance @divyansh42 with the labeling between

https://github.com/redhat-appstudio/o11y/blob/beee5d1ad74a1db6de2250a5e32096f38ee7e052/test/promql/tests/data_plane/pipeline_deadlock_crashloop_test.yaml#L271

and 

https://github.com/redhat-appstudio/o11y/blob/beee5d1ad74a1db6de2250a5e32096f38ee7e052/test/promql/tests/data_plane/pipeline_deadlock_crashloop_test.yaml#L269

See if you can figure it out during the earlier part of your day on Monday, and then we'll sync up per usual on Monday morning for me / later for you

File: test/promql/tests/data_plane/pipeline_overhead_test.yaml
Comment (gabemontero) [inline test/promql/tests/data_plane/pipeline_overhead_test.yaml:263]:
```suggestion
      - series: 'pac_watcher_client_latency_bucket{le="1", source_cluster="cluster02"}'
```

similar labeling nuanced diff when you copy/pasted from the Chains alert

File: test/promql/tests/data_plane/pipeline_overhead_test.yaml
Comment (gabemontero) [inline test/promql/tests/data_plane/pipeline_overhead_test.yaml:265]:
```suggestion
      - series: 'pac_watcher_client_latency_bucket{le="+Inf", source_cluster="cluster02"}'
```

again labeling nuance

File: test/promql/tests/data_plane/pipeline_overhead_test.yaml
Comment (gabemontero) [inline test/promql/tests/data_plane/pipeline_overhead_test.yaml:283]:
```suggestion
              runbook_url: https://gitlab.cee.redhat.com/konflux/docs/sop/-/blob/main/pipeline-service/slos/pac-performance.md #TODO: change link
```

self explanatory I would think

**Review by gabemontero — CHANGES_REQUESTED** [review]
With these suggested changes to your branch I as able to run the `make check-and-test` on my checkout of your branch @divyansh42 

```
<truncated first portion>
INFO[0004] [test/promql/tests/data_plane/persistentvolumeissues_test.yaml] done 
INFO[0005] [test/promql/tests/data_plane/pipeline_deadlock_crashloop_test.yaml] done 
INFO[0011] [test/promql/tests/data_plane/pipeline_overhead_test.yaml] done 
INFO[0011] [test/promql/tests/data_plane/quota_exceeded_test.yaml] done 
INFO[0011] [test/promql/tests/data_plane/release_service_test.yaml] done 
INFO[0011] [test/promql/tests/data_plane/unschedulable_pods_test.yaml] done 
INFO[0011] ALL DONE OK! :-)                             
/usr/bin/podman run -v "/home/gmontero/go/src/github.com/redhat-appstudio/o11y:/work" --rm --privileged -t quay.io/rhobs/obsctl-reloader-rules-checker:1.0.4 -t rhtap -t rhtap -d rhobs/recording -y -p --tests-dir test/promql/tests/recording
INFO[0000] checking rules...                            
INFO[0000] [rhobs/recording/build_service_availability_recording_rules.yaml] done 
INFO[0000] [rhobs/recording/exporter_recording_rules.yml] done 
INFO[0000] [rhobs/recording/image_controller_availability_recording_rules.yaml] done 
INFO[0000] [rhobs/recording/integration_service_availability_recording_rules.yaml] done 
INFO[0000] [rhobs/recording/release_service_availability_recording_rules.yaml] done 
INFO[0000] running YAML linter...                       
INFO[0000] [rhobs/recording/build_service_availability_recording_rules.yaml] done 
INFO[0000] [rhobs/recording/exporter_recording_rules.yml] done 
INFO[0000] [rhobs/recording/image_controller_availability_recording_rules.yaml] done 
INFO[0000] [rhobs/recording/integration_service_availability_recording_rules.yaml] done 
INFO[0000] [rhobs/recording/release_service_availability_recording_rules.yaml] done 
INFO[0000] [test/promql/tests/recording/build_service_availability_recording_rules_test.yaml] done 
INFO[0000] [test/promql/tests/recording/exporter_recording_rules_test.yml] done 
INFO[0001] [test/promql/tests/recording/image_controller_availability_recording_rules_test.yaml] done 
INFO[0001] [test/promql/tests/recording/integration_service_availability_recording_rules_test.yaml] done 
INFO[0001] [test/promql/tests/recording/release_service_availability_recording_rules_test.yaml] done 
INFO[0001] running 'pint' linter...                     
INFO[0001] [rhobs/recording/build_service_availability_recording_rules.yaml] done 
INFO[0001] [rhobs/recording/exporter_recording_rules.yml] done 
INFO[0001] [rhobs/recording/image_controller_availability_recording_rules.yaml] done 
INFO[0001] [rhobs/recording/integration_service_availability_recording_rules.yaml] done 
INFO[0001] [rhobs/recording/release_service_availability_recording_rules.yaml] done 
INFO[0001] running the unit tests...                    
INFO[0001] [test/promql/tests/recording/build_service_availability_recording_rules_test.yaml] done 
INFO[0001] [test/promql/tests/recording/exporter_recording_rules_test.yml] done 
INFO[0001] [test/promql/tests/recording/image_controller_availability_recording_rules_test.yaml] done 
INFO[0001] [test/promql/tests/recording/integration_service_availability_recording_rules_test.yaml] done 
INFO[0001] [test/promql/tests/recording/release_service_availability_recording_rules_test.yaml] done 
INFO[0001] ALL DONE OK! :-)                             
gmontero ~/go/src/github.com/redhat-appstudio/o11y  (local-upstream-349-copy)$ 
```

See if this helps you in the earlier part of your Monday while I am sleeping, and then we'll chat per usual my Monday AM your Monday PM.

**Comment by divyansh42** [comment]
@gabemontero will post screenshot for the metrics regex trial on OCP console soon.

File: test/promql/tests/data_plane/pipeline_deadlock_crashloop_test.yaml
Comment (gabemontero) [inline test/promql/tests/data_plane/pipeline_deadlock_crashloop_test.yaml:309]:
```suggestion
      - series: 'pac_watcher_work_queue_depth{source_cluster="cluster01"}'
```

File: test/promql/tests/data_plane/pipeline_deadlock_crashloop_test.yaml
Comment (gabemontero) [inline test/promql/tests/data_plane/pipeline_deadlock_crashloop_test.yaml:321]:
```suggestion
              source_cluster: cluster01
```

File: test/promql/tests/data_plane/pipeline_deadlock_crashloop_test.yaml
Comment (gabemontero) [inline test/promql/tests/data_plane/pipeline_deadlock_crashloop_test.yaml:326]:
```suggestion
                Tekton PAC controller on cluster cluster01 has appeared deadlocked on 600 pipelinerun events.
```

**Review by gabemontero — CHANGES_REQUESTED** [review]
are you running `make check-and-test` in your local fork before pushing commits @divyansh42 ?

With the crashloop test, you have both cluster labels as `cluster02` and so the alert did not fire.  I've provided suggested changes that allow that test to pass for me locally.

Then, while you added the test for the `PACControllerPerformance` alert, you did not actually add the alert in the alert file.

I'll leave that to you to try out with you next push.

**Comment by divyansh42** [comment]
> Then, while you added the test for the PACControllerPerformance alert, you did not actually add the alert in the alert file.
I have removed that alert as there was conversation going around removing the performance alerts

**Comment by gabemontero** [comment]
> > Then, while you added the test for the PACControllerPerformance alert, you did not actually add the alert in the alert file.
> 
> I have removed that alert as there was conversation going around removing the performance alerts. Please do let me know if performance alert need to be added. Otherwise, I will remove the test file as well.

You are correct about the conversation but the final call has not been made yet.  Go ahead and try to add the alert and get the test going, testing locally with `make check-and-test` .... it will be good practice for you.  We can remove later in the week, or even remove later in this PR, if need be.

**Comment by divyansh42** [comment]
@gabemontero I have pushed the changes but `make check-and-test` is still failing.
We can discuss that further in today's call.

**Comment by gabemontero** [comment]
squash your commits and maitain the chore(SRVKP-4528)... text in the remaining commit, and make sure the commit is signed @divyansh42  ... then the git-lint failure should go away

File: rhobs/alerting/data_plane/prometheus.pipeline_alerts.yaml
Comment (gabemontero) [inline rhobs/alerting/data_plane/prometheus.pipeline_alerts.yaml:52]: I tried the query `kube_pod_container_status_restarts_total{namespace="openshift-pipelines", pod=~"tekton-.*|pipelines-as-code-.*"}` in the prod-rh01 OCP metrics console and it looks correct @divyansh42 

Good job figuring out the regex.

File: rhobs/alerting/data_plane/prometheus.pipeline_alerts.yaml
Comment (gabemontero) [inline rhobs/alerting/data_plane/prometheus.pipeline_alerts.yaml:167]:
```suggestion
            (sum by (source_cluster) (increase(pac_watcher_client_latency_bucket{le='10'}[5m]))
```

I have a PR up to change this for chains as well, so don't worry, just an adjustment on my part

**Review by gabemontero — CHANGES_REQUESTED** [review]
just one minor tweak @divyansh42 that is on me

pull that in, clean up the commits, and we'll start tagging o11y and app sre

I'm removing the WIP prefix.

**Comment by gabemontero** [comment]
> just one minor tweak @divyansh42 that is on me
> 
> pull that in, clean up the commits, and we'll start tagging o11y and app sre
> 
> I'm removing the WIP prefix.

I don't have permission to remove the WIP prefix for PRs in this repo ... you'll have to update the PR title @divyansh42

File: rhobs/alerting/data_plane/prometheus.pipeline_alerts.yaml
Comment (divyansh42) [reply rhobs/alerting/data_plane/prometheus.pipeline_alerts.yaml:167]: So, do I need to do the changes? or will you be covering that in your chains PR?

**Comment by divyansh42** [comment]
> pull that in, clean up the commits, and we'll start tagging o11y and app are

Done.

**Comment by gabemontero** [comment]
@amisstea @redhat-appstudio/konflux-sre PTAL

Now, after approvals, we'll have to figure out which order we merge PRs in @amisstea .... my proposal is : 
- https://github.com/redhat-appstudio/o11y/pull/342
- https://github.com/redhat-appstudio/o11y/pull/354
- https://github.com/redhat-appstudio/o11y/pull/350
- then this PR

thanks

**Review by amisstea — APPROVED** [review]

**Review by amisstea — CHANGES_REQUESTED** [review]
@gabemontero please fix the merge conflicts.

**Comment by divyansh42** [comment]
Rebased and tried to fix the gridPos as well.
@gabemontero please take a look. Thanks!

**Comment by gabemontero** [comment]
the yaml lint failed @divyansh42 
```
time="2024-08-28T20:21:16Z" level=info msg="[rhobs/alerting/data_plane/prometheus.latency_snapshot_to_static_integration_plr_creation_alerts.yaml] done\n"
time="2024-08-28T20:21:16Z" level=fatal msg="[rhobs/alerting/data_plane/prometheus.pipeline_alerts.yaml] does not store a Kube object serialized in YAML: yaml: line 8: did not find expected key\n"
make: *** [Makefile:23: check-and-test] Error 1

```

there is a `make lint_yamls` that if you have python3 could let you debug locally on Thursday

I'll check in with you my Thursday AM and we'll see where you are at and decide the most expedient path

**Comment by divyansh42** [comment]
> the yaml lint failed @divyansh42

It should not fail now. I accidentally removed one line in prometheus.pipeline_alerts.yaml

File: dashboards/grafana-dashboard-rhtap-slos.configmap.yaml
Comment (gabemontero) [inline dashboards/grafana-dashboard-rhtap-slos.configmap.yaml:2159]: yeah so looking at the current main branch @amisstea @divyansh42 the y axis here seem sufficiently below https://github.com/redhat-appstudio/o11y/blob/dc205d9cdd390e0502d6367cf1949cc3a547892f/dashboards/grafana-dashboard-rhtap-slos.configmap.yaml#L2111-L2115 on the y axis that this should render somewhere below the last panel in main branch at this time.

I say we merge this, and then I can do the one time fix to group the rows correctly, remove any gaps, move the sample panels back to the bottom, etc. on Thursday.

Comment (divyansh42) [reply dashboards/grafana-dashboard-rhtap-slos.configmap.yaml:2159]: I have tried to adjust it to appear at the bottom of the chains panel.

Comment (gabemontero) [reply dashboards/grafana-dashboard-rhtap-slos.configmap.yaml:2159]: Yep that is what I was detailing in my last comment ... thanks for providing the clarifying summary @divyansh42 

let's wait for @amisstea to log in and get to this, and then we can see how it looks in https://grafana.stage.devshift.net/d/rhtap-slos/rhtap-slos?orgId=1&refresh=1h&from=now-5m&to=now

**Comment by gabemontero** [comment]
> @gabemontero please fix the merge conflicts.

yeah @amisstea see my comment https://github.com/redhat-appstudio/o11y/pull/349#discussion_r1735400158 ... I think this is ready to merge if you are OK

also, I don't have permission in this repo to rebase other people's PRs

**Review by amisstea — APPROVED** [review]

File: dashboards/grafana-dashboard-rhtap-slos.configmap.yaml
Comment (amisstea) [reply dashboards/grafana-dashboard-rhtap-slos.configmap.yaml:2159]: I'm good with merging this and fixing the positioning in a followup PR.

**Comment by gabemontero** [comment]
thanks @amisstea nothing grossly askew with  https://grafana.stage.devshift.net/d/rhtap-slos/rhtap-slos?orgId=1&refresh=1h&from=now-5m&to=now and it has @divyansh42 's PAC panels

I'll start working on the ordering tweaks we talked about and get a PR up later today

**Pull request merged**
